### PR TITLE
Make MostInlined and BestEffort inline propagation no longer assert replayed

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Currently, when doing `MostInlined` propagation, at each step `propagateX2Y`, the inline propagator `TORCH_CHECK` that Y is already replayed consistently with X at -1, then it will inline Y with the innermost dimension of X. `BestEffort` is similar except that it allows specifying a position.

When writing my transpose PR https://github.com/csarofeen/pytorch/pull/1854, I don't feel this behavior very convenient to use. For example, if I have a DAG:

```
T0[I1, I2, I3] = input
T1[I1, I2, I3] = sin(T0)
T2[I1, I2, I3] = cos(T1)
T3[I1, I3, I2] = transpose(T2)
T4[I1, I3, I2] = exp(T3)
output = T5[I1, I3, I2] = tan(T4)
```

If I do a `MostInlined` inline propagation starting from `T5`, I will get an error message when propagating from `T3` to `T2` saying that dims do not match and replay is required. But what I really want is to just don't inline into the mismatched dimension and continue the propagation so that the `sin` will be inlined into the innermost loop of `cos`.

I think generally, it makes sense to make `MostInlined` and `BestEffort` to just accept the current replay status as is, and just don't inline into mismatched dimensions.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Making this a draft to play with something and add unit tests.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/inline_propagator.cpp [hunk 1]
```diff
@@ -126,7 +126,23 @@ size_t MaxPosCalculator::getMaxProducerPosFromConsumer(
...
+  size_t max_producer_pos = 0;
+  for (int64_t consumer_pos = consumer->nDims(); consumer_pos >= 0;
+       consumer_pos--) {
+    auto producer_pos = TransformReplay::getMatchedLeafPosWithoutReplayPasC(
+        producer, consumer, consumer_pos);
+    if (producer_pos >= 0) {
+      max_producer_pos = producer_pos;
+      break;
+    }
+  }
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/inline_propagator.cpp]: Note to myself: clean this up when https://github.com/csarofeen/pytorch/pull/1870 is merged

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/inline_propagator.cpp]: rebased and cleaned up

**Comment by zasdfgbnm** [comment]
Rebased to https://github.com/csarofeen/pytorch/pull/1871

**Pull request merged**